src/value/native_reader.rs

```diff
@@ -0,0 +1,125 @@
+use crate::raw_reader::RawReader;
+use crate::result::IonResult;
+use crate::text::raw_text_reader::RawTextReader;
+use crate::value::owned;
+use crate::value::owned::{OwnedElement, OwnedSequence, OwnedStruct, OwnedValue};
+use crate::value::reader::ElementReader;
+use crate::{IonType, RawBinaryReader, Reader, StreamItem, StreamReader};
+
+/// Provides an implementation of [ElementReader] that is backed by a native Rust [Reader].
+pub struct NativeElementReader;
+
+struct NativeElementIterator<R: RawReader> {
+    reader: Reader<R>,
+}
+
+impl<R: RawReader> Iterator for NativeElementIterator<R> {
+    type Item = IonResult<OwnedElement>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.materialize_next().transpose()
+    }
+}
+
+impl ElementReader for NativeElementReader {
+    fn iterate_over<'a, 'b>(
+        &'a self,
+        data: &'b [u8],
+    ) -> IonResult<Box<dyn Iterator<Item = IonResult<OwnedElement>> + 'b>> {
+        // Match against the slice to see if it starts with the binary IVM
+        match data {
+            &[0xe0, 0x01, 0x00, 0xea, ..] => {
```

Comment:
Is there any way to get re-use here between this format detection and the IVM parsing elsewhere in the library?